File: aoc23/day1/day1.py


import sys


res1=0
res2=0

for line in sys.stdin:

    l=0;
    r=0;
    for c in line:
        if c >= '0' and c <= '9':
            if (l==0):
                l=ord(c)-ord('0');
            r=ord(c)-ord('0');
    
    res1+= l*10+r

    #######

    l=0;
    r=0;

    num={1:"one", 2:"two", 3:"three", 4:"four", 5:"five", 6:"six", 7:"seven", 8:"eight",9:"nine"}

    for i in range(len(line)):
        for key in num:
            c = line[i]
            if c >= '0' and c <= '9':
                if (l==0):
                    l=ord(c)-ord('0');
                r=ord(c)-ord('0');
            elif line[i:].startswith(num[key]):
                if l == 0:
                    l=key
                r=key
    res2+= l*10+r


if len(sys.argv) > 1 and sys.argv[1] == '2':
    print(res2);
else:
    print(res1);

File: aoc23/day1/day1.nim
import std/strutils
# import prelude

var
    first: char = ' '
    last: char
    sum: int

    firstp2: int = 0
    lastp2: int
    sump2: int

const nbStringArray = ["one", "two", "three", "four", "five", "six", "seven", "eight", "nine"]

proc checkNb(line: string, i: int): int =
    if isDigit(line[i]):
        return ord(line[i]) - ord('0')
    for nbMapIndex in 0 ..< nbStringArray.len:
        if line.continuesWith(nbStringArray[nbMapIndex], i):
            return nbMapIndex + 1 
        
    return 0



for line in stdin.lines:

    first = ' '
    last = ' '
    for ch in line:
        if isDigit(ch):
            if first == ' ':
                first = ch
            last = ch

    sum += (ord(first) - ord('0')) * 10 + (ord(last) - ord('0'))

    firstp2 = 0
    lastp2 = 0
    for i in 0 ..< len(line):
        let nb = checkNb(line, i)
        if nb > 0:  
            if firstp2 == 0:
                firstp2 = nb
            lastp2 = nb
    
    sump2 += firstp2 * 10 + lastp2
    
echo "part1 ", sum
    
echo "part2 ", sump2